src/libPersistence/IncrementalDB.cpp

```diff
@@ -0,0 +1,465 @@
+                                    const string& dbName) {
+  if (m_DBPointer.at(dbName).first != currentDSEpoch) {
+    m_DBPointer.at(dbName).first = currentDSEpoch;
+    m_DBPointer.at(dbName).second =
+        make_shared<LevelDB>(dbName, m_path, to_string(currentDSEpoch));
+  }
+}
+
+bool IncrementalDB::PutTxBody(const dev::h256& txID, const bytes& body,
+                              const uint64_t& dsEpoch) {
+  ChangeDBPointer(dsEpoch, m_txBodyDBName);
+
+  int ret = -1;
+
+  ret = m_DBPointer.at(m_txBodyDBName).second->Insert(txID, body);
+
+  return (ret == 0);
+}
+
+bool IncrementalDB::PutMicroBlock(const BlockHash& blockHash, const bytes& body,
+                                  const uint64_t& dsEpoch) {
+  ChangeDBPointer(dsEpoch, m_microBlockDBName);
+
+  int ret = -1;
+
+  ret = m_DBPointer.at(m_microBlockDBName).second->Insert(blockHash, body);
+
+  return (ret == 0);
+}
+
+bool IncrementalDB::PutTxBlock(const uint64_t& blockNum, const bytes& body,
+                               const uint64_t& dsEpoch) {
+  ChangeDBPointer(dsEpoch, m_TxBlockDBName);
+
+  int ret = -1;
+
+  ret = m_DBPointer.at(m_TxBlockDBName).second->Insert(blockNum, body);
+
+  return (ret == 0);
+}
+
+bool IncrementalDB::PutDSBlock(const uint64_t& blockNum, const bytes& body,
+                               const uint64_t& dsEpoch) {
+  ChangeDBPointer(dsEpoch, m_DSBlockDBName);
+
+  int ret = -1;
+
+  ret = m_DBPointer.at(m_DSBlockDBName).second->Insert(blockNum, body);
+
+  return (ret == 0);
+}
+
+bool IncrementalDB::PutFallbackBlock(const BlockHash& blockHash,
+                                     const bytes& body,
+                                     const uint64_t& dsEpoch) {
+  ChangeDBPointer(dsEpoch, m_FallbackBlockDBName);
+
+  int ret = -1;
+
+  ret = m_DBPointer.at(m_FallbackBlockDBName).second->Insert(blockHash, body);
+
+  return (ret == 0);
+}
+
+bool IncrementalDB::PutVCBlock(const BlockHash& blockHash, const bytes& body,
+                               const uint64_t& dsEpoch) {
+  ChangeDBPointer(dsEpoch, m_VCBlockDBName);
+
+  int ret = -1;
+
+  ret = m_DBPointer.at(m_VCBlockDBName).second->Insert(blockHash, body);
+
+  return (ret == 0);
+}
+
+bool IncrementalDB::PutBlockLink(const uint64_t& index, const bytes& body) {
+  int ret = -1;
+  ret = m_blockLinkDB->Insert(index, body);
+  return (ret == 0);
+}
+
+void IncrementalDB::Init() {
+  const std::string path_abs = "./" + m_path;
+  if (!boost::filesystem::exists(path_abs)) {
+    boost::filesystem::create_directories(path_abs);
+  }
+
+  string emptyString;
+  emptyString.clear();
+  m_blockLinkDB = make_shared<LevelDB>(m_blockLinkDBName, m_path, emptyString);
+
+  for (auto const& dbName :
+       {m_txBodyDBName, m_microBlockDBName, m_VCBlockDBName, m_DSBlockDBName,
+        m_FallbackBlockDBName, m_TxBlockDBName}) {
+    m_DBPointer.emplace(
+        dbName,
+        make_pair(0, make_shared<LevelDB>(dbName, m_path, (string) "0")));
+  }
```

Comment:
is it necessary to initialize the DS 0 db explicitly?

Reply:
yes, otherwise the first call will cause segmentation fault.